### PR TITLE
33 login adjustments

### LINKED ISSUE #31: Add logout
Add logout functionality that revokes permissions until the next login.

### LINKED ISSUE #33: Adjusting the login structure
We should make a distinction between what functionalities are available to users that are logged in vs users that are not logged in.

We could group these into a login page and a user dashboard/menu.  

The login page should contain:
- the login form 
- buttons that redirect to the register user form
- buttons that redirect to the password reset form (I forgot my password) 

The a user dashboard/menu page could be the already existing members homepage which is already login protected. This one should contain buttons for:
- Redirecting to change password form
- Logout
This means the change password form should also be login protected, and since only a logged in user can access it, the username in the form and back-end authentication would no longer be required.

### LINKED ISSUE #35: Reject logins for users with `active=False`
Adjust the authentication such that if the `active` field in the members table is set to `False`,  the user is not granted access to the system. This would an account suspended scenario.

### LINKED ISSUE #44: Validate passwords on the server rather than the database
When reviewing #42 I was able to create an account with a weak password. I think this is due to the fact that the password is validate with the db model and at that point it's hashed. So we're essentially validating the strength of the hash rather than the password.

### PR DESCRIPTION
**Pull request by iiacoban42**
Added the logout functionality, change password form is login protected, only active users can login and the password is now validated on the server.

Closes #33
Closes #31
Closes #35
Closes #44

### CODE REVIEW

**Comment by StefanoSperetta** [comment]
Can you also add the allowed characters for a password to the registration webpage so users know what they are supposed to do?

File: src/members/models.py [hunk 1]
```diff
@@ -10,19 +9,9 @@ class Member(AbstractUser):
...
-                                validators=[
-                                    RegexValidator(
-                                        regex='^(?=.*[a-z])(?=.*[A-Z])'
-                                              '(?=.*\\d)(?=.*[@$!%*?&])'
-                                              '[A-Za-z\\d@$!%*?&]{8,}$',
-                                        message='Enter password containing min 8 characters, '
-                                                'min 1 upper and lower case letter, '
-                                                '1 number and 1 special character')
-                                ], null=True, blank=True)
+    password = models.CharField(max_length=120, null=True, blank=True)
```
Comment (NotUnlikeTheWaves) [inline src/members/models.py:12]: Should the password field be nullable?

Comment (iiacoban42) [reply src/members/models.py]: It should not, good that you caught that!

**Comment by StefanoSperetta** [comment]
One more comment on the password / form text. Special character (https://www.webopedia.com/definitions/special-character/ I guess this is a complete list) is quite generic and only a part of those special characters are supported ( @ $ ! % * ? & ).

I would add the list of special characters supported to the form (...1 lower case letter, 1 number and 1 special character (@ $ ! % * ? &).

Could you add a couple of extra characters like - or _? (this is a personal preference, happy to get suggestions here ;-)) The regexp would become: ^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[@$!%*?&_\-])[A-Za-z\\d@$!%*?&_\-]{8,}$

**Comment by NotUnlikeTheWaves** [comment]
Alternatively we could go for a much simpler check with a few if statements. Very rough psuedocode:
```py
if password.AnyUppercase and password.AnyLowercase and password.AnyNumber and password.AnyNonAlphaNumeric:
    # proceed
else:
    # return error
```

This avoids unreadable regexes in the first place and possibly also gives us an error message as to which part of the password is missing

Also as a general question, what's the reason for the required password complexity? We for example would immediately disallow passphrases (https://xkcd.com/936/)

**Comment by StefanoSperetta** [comment]
Very good point... One of the reasons was to have a "solid" password and prevent some silly SQL injection attacks (this might be an example https://realpython.com/prevent-python-sql-injection/ ). In general, some sanitization would be good and the regexp was achieving that purpose. But I agree it might be best to re-think the problem and eventually just force a certain length and sanitize the password / username.

**Comment by iiacoban42** [comment]
Shall we use Django's password validation? Seems to be quite strong without that many constraints. 
```
AUTH_PASSWORD_VALIDATORS = [
    {
        'NAME': 'django.contrib.auth.password_validation.UserAttributeSimilarityValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.MinimumLengthValidator',
        'OPTIONS': {
            'min_length': 9,
        }
    },
    {
        'NAME': 'django.contrib.auth.password_validation.CommonPasswordValidator',
    },
    {
        'NAME': 'django.contrib.auth.password_validation.NumericPasswordValidator',
    },
]
```
image

**Comment by StefanoSperetta** [comment]
Sounds like a good plan!

**Comment by iiacoban42** [comment]
Alright, done!

**Comment by StefanoSperetta** [comment]
Fixed a typo that was causing the username to always be set to the email address.

**Pull request merged**